crates/ruff_server/src/server/api/requests/execute_command.rs

```diff
@@ -0,0 +1,144 @@
+use std::collections::HashMap;
+
+use crate::server;
+use crate::server::api::LSPResult;
+use crate::server::client;
+use crate::server::schedule::Task;
+use crate::session::Session;
+use crate::DIAGNOSTIC_NAME;
+use lsp_server::ErrorCode;
+use lsp_types::{self as types, request as req};
+use serde::Deserialize;
+
+enum Command {
+    Format,
+    FixAll,
+    OrganizeImports,
+}
+
+pub(crate) struct ExecuteCommand;
+
+#[derive(Deserialize)]
+struct TextDocumentArgument {
+    uri: types::Url,
+}
+
+impl super::RequestHandler for ExecuteCommand {
+    type RequestType = req::ExecuteCommand;
+}
+
+impl super::SyncRequestHandler for ExecuteCommand {
+    fn run(
+        session: &mut Session,
+        _notifier: client::Notifier,
+        requester: &mut client::Requester,
+        params: types::ExecuteCommandParams,
+    ) -> server::Result<Option<serde_json::Value>> {
+        let Some(command) = Command::from_str(&params.command) else {
+            return Err(anyhow::anyhow!("")).with_failure_code(ErrorCode::InvalidParams);
+        };
+
+        let mut changes = HashMap::new();
+        let documents =
+            args_as_text_documents(params.arguments).with_failure_code(ErrorCode::InvalidParams)?;
```

Comment:
Can you confirm whether `changes` support having multiple documents? I can't exactly figure out reading the spec but I presume it does. But, if it doesn't then this will be problematic for Jupyter Notebooks where each cell is its own text document so the edits are tied to multiple documents.

Reply:
`changes` does enable multi-document edits, though none of the commands will modify multiple files. I think the LSP spec allows treating notebook files as a single document, so we could still have a single `changes` entry that covers multiple cells in a notebook.

Reply:
> I think the LSP spec allows treating notebook files as a single document, so we could still have a single `changes` entry that covers multiple cells in a notebook.

Yes, but it's not about a Notebook document here. How would the client know which _cell_ does this edit belong to? This is only known through the text document URI because each cell is represented using text document. Refer to [this loop](https://github.com/astral-sh/ruff-lsp/blob/187d7790be0783b9ac41ce025a724cf389bf575c/ruff_lsp/server.py#L1352-L1367) which does that in `ruff-lsp`.